cep-purls.md

```diff
@@ -0,0 +1,192 @@
+```
+
+Tools that build conda packages should add these packages to `index.json`:
+
+```json
+{
+  ...,
+  "purls": ["pkg:pypi/pinject@0.14.1"]
+}
+```
+
+### PURLs in repodata
+
+In `repodata.json`, PURLs should be an optional field.
+Tools that generate `repodata.json` (like `rattler-index` or `conda-index`) can decide whether to include it or not in `repodata.json`.
+
+Instead of writing the PURLs in `repodata.json`, these tools can decide to add PURLs to a separate `purls.json` like in [CEP 12](cep-0012.md):
+
+```json
+{
+  "info": {
+    "platform": "string",
+    "arch": "string",
+    "subdir": "string",
+    "version": 0
+  },
+  "packages": {
+    "package-version-build.tar.bz2": {
+      ...,
+      "purls": ["pkg:pypi/pinject@0.14.1"]
+    }
+  },
+  "packages.conda": {
+    "package-version-build.conda": {
+      ...,
+      "purls": ["pkg:pypi/pinject@0.14.1"]
+    }
+  }
+}
+```
+
+If the `purls` field is not present in the `repodata.json` record, it means no `purls` information is stored with the record. Then, a fallback mechanism (i.e., read `purls.json`) should be used to acquire the PURL information.
+
+In sharded repodata ([CEP 16](cep-0016.md)), PURLs can be specified similar to [CEP 21](cep-0021.md).
+
+### Patching
+
+With the purls part of the repodata we propose to also allow repodata patching this field. The original purls can still be extracted from the packages in `index.json`. We do not see a reason how the PURL information is different from other patchable information stored in the repodata (like the dependencies).
+
+To facilitate these patches, the `patch_instructions_version` in `patch_instructions.json` files is incremented to `3`. A `patch_instructions_version: 3` file MAY contain a `purls: ...` field in the patch instructions that MUST be used to patch generated the `purls` entry in the (sharded) repodata (or the `purls.json`).
+
+```json
+{
+  "patch_instructions_version": 3,
+  "packages": {
+    "package-version-build.tar.bz2": {
+      ...,
+      "purls": ["pkg:pypi/pinject@0.14.1"]
+    },
+    ...
+  }
+}
+```
+
+### PURL of a conda packages
+
+Conda packages itself should have a PURL as well.
+This makes it possible for CVE Numbering Autorities (CNA) to publish vulnerabilities for conda packages.
+A package `pinject-0.14.1-pyh29332c3_1.conda` published on `conda-forge` should have the PURL `pkg:conda/conda-forge/pinject@0.14.1?build=pyh29332c3_1&platform=noarch`.
+Packages can also specify a custom `repository_url`, for example `pkg:conda/custom-channel/my-package@0.1.0?build=pyh29332c3_1&platform=noarch&repository_url=prefix.dev`.
+
+Tools that generate packages like rattler-build and conda-build should also be able to inject a PURL at build time via CLI flags.
+For this, v1-recipes can use the jinja syntax from the recipes.
+
+```bash
+rattler-build build -r recipe/ \
+    --append-purl-pattern \
+    'pkg:conda/conda-forge/${{ PACKAGE_NAME }}@${{ PACKAGE_VERSION }}?build=${{ BUILD_NUMBER }}'
```

Comment:
I don't see why we need to inject the PURL of the package being built into itself. All this information can be derived from the filename if we want to publish it somewhere else, but I don't see how publishing an easily buildable string adds value.

Also, the `build` qualifier is the build _string_ not the number.